core/src/main/java/com/linecorp/armeria/server/RoutingDecorator.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.function.Function;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+
+import io.netty.util.AttributeKey;
+
+/**
+ * A pair of a {@link Route} and a {@code decorator} bound to it.
+ */
+final class RoutingDecorator {
```

Comment:
How about making this a real decorator? i.e. `RoutingDecorator` could be merged with `DynamicDispatcherService`.

We could then rename this to `RouteDecoratingService`?